os/win/dnd.cpp

```diff
@@ -54,129 +54,139 @@ gfx::Point drag_position(HWND hwnd, POINTL& pt)
+public:
+  DataWrapper() = delete;
+  DataWrapper(const DataWrapper&) = delete;
+  DataWrapper(IDataObject* data) : m_data(data) {}
+
+  ~DataWrapper() { release(); }
+
+  template<typename T>
+  GLocked<T> get(CLIPFORMAT cfmt) {
+    release();
+
+    FORMATETC fmt;
+    fmt.cfFormat = cfmt;
+    fmt.ptd = nullptr;
+    fmt.dwAspect = DVASPECT_CONTENT;
+    fmt.lindex = -1;
+    fmt.tymed = TYMED::TYMED_HGLOBAL;
+    m_result = m_data->GetData(&fmt, &m_medium);
+    if (m_result != S_OK)
+      return nullptr;
+
+    return GLocked<T>(m_medium.hGlobal);
+  }
+
+private:
+  void release() {
+    if (m_result == S_OK) {
+      ReleaseStgMedium(&m_medium);
+      m_result = -1L;
+    }
+  }
+
+  IDataObject* m_data = nullptr;
+  STGMEDIUM m_medium;
+  HRESULT m_result = -1L;
+};
+
 } // anonymous namespace
 
 namespace os {
 
 base::paths DragDataProviderWin::getPaths()
 {
   base::paths files;
-  FORMATETC fmt;
-  fmt.cfFormat = CF_HDROP;
-  fmt.ptd = nullptr;
-  fmt.dwAspect = DVASPECT_CONTENT;
-  fmt.lindex = -1;
-  fmt.tymed = TYMED::TYMED_HGLOBAL;
-  STGMEDIUM medium;
-  if (m_data->GetData(&fmt, &medium) == S_OK) {
-    {
-      GLock<HDROP> hdrop(medium.hGlobal);
-      if (static_cast<HDROP>(hdrop)) {
-        int count = DragQueryFile(hdrop, 0xFFFFFFFF, nullptr, 0);
-        for (int index = 0; index < count; ++index) {
-          int length = DragQueryFile(hdrop, index, nullptr, 0);
-          if (length > 0) {
-            // From Win32 docs: https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-dragqueryfilew
-            // the DragQueryFile() does't include the null character in its return value.
-            std::vector<TCHAR> str(length + 1);
-            DragQueryFile(hdrop, index, str.data(), str.size());
-            files.push_back(base::to_utf8(str.data()));
-          }
-        }
+  DataWrapper data(m_data);
+  GLocked<HDROP> hdrop = data.get<HDROP>(CF_HDROP);
+  if (hdrop != nullptr) {
+    int count = DragQueryFile(hdrop, 0xFFFFFFFF, nullptr, 0);
+    for (int index = 0; index < count; ++index) {
+      int length = DragQueryFile(hdrop, index, nullptr, 0);
+      if (length > 0) {
+        // From Win32 docs: https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-dragqueryfilew
+        // the DragQueryFile() doesn't include the null character in its return value.
+        std::vector<TCHAR> str(length + 1);
+        DragQueryFile(hdrop, index, str.data(), str.size());
+        files.push_back(base::to_utf8(str.data()));
       }
     }
-    ReleaseStgMedium(&medium);
   }
   return files;
 }
 
 SurfaceRef DragDataProviderWin::getImage()
 {
   SurfaceRef surface = nullptr;
-  clip::image_spec spec;
   clip::image img;
-  SurfaceFormatData sfd;
-
-  STGMEDIUM medium;
-  medium.hGlobal = nullptr;
-  FORMATETC fmt;
-  fmt.ptd = nullptr;
-  fmt.dwAspect = DVASPECT_CONTENT;
-  fmt.lindex = -1;
-  fmt.tymed = TYMED_HGLOBAL;
 
+  auto makeSurface = [](const clip::image& img) {
```

Comment:
I know this could be a private function, but I preferred to do it here which is closer to the place it is used. Also, this is going to disappear in a future commit since os::instance()->makeSurface() will be called directly after its pending changes are implemented.